scripts/opentitan/autoreg.py

```diff
@@ -0,0 +1,953 @@
+        else:
+            noaccess_types = (QEMUAccess.WO, )
+        noaccess_names = {
+            r.name for r in regs if r.access in noaccess_types
+        }
+        rc_names = {
+            r.name for r in regs if r.access == QEMUAccess.RC
+        }
+        reg_names -= noaccess_names
+        if not write and rc_names:
+            reg_names -= rc_names
+        rwidth = self._regwidth
+        dname =self._devname
+        hdname = dname.title().replace('_', '')
+        vnm = f'val{rwidth}'
+        if write:
+            if rwidth < 64:
+                vcast = f'uint{rwidth} val{rwidth} = (uint{rwidth})val64;'
+            else:
+                vcast = ''
+            code = f'''
+            static void {dname}_regs_write(void *opaque, hwaddr addr,
+                                           uint64_t val64, unsigned size)
+            {{
+                {hdname}State *s = opaque;
+                (void)size;
+                {vcast}
+                hwaddr reg = R{rwidth}_OFF(addr);
+
+                uint32_t pc = ibex_get_current_pc();
+                trace_{dname}_io_write(s->ot_id, (uint32_t)addr, REG_NAME(reg),
+                                       val{rwidth}, pc);
+            '''
+        else:
+            if rwidth < 64:
+                vdef = f'uint{rwidth} val{rwidth};'
+            else:
+                vdef = ''
+            code = f'''
+            static uint64_t {dname}_regs_read(void *opaque, hwaddr addr,
+                                              unsigned size)
+            {{
+                {hdname}State *s = opaque;
+                (void)size;
+                {vdef}
+
+                hwaddr reg = R{rwidth}_OFF(addr);
+            '''
+        print(self.redent(code), file=tfp)
+        lines = []
+        lines.append('    switch (reg) {')
+        bitfield = (1 << rwidth) - 1
+        if not write:
+            for rname in sorted(reg_names, key=lambda r: nregs[r].address):
+                lines.append(f'case R_{rname}:')
+            lines.append(f'    {vnm} = s->regs[reg];')
+            lines.append('    break;')
+            for rname in sorted(rc_names, key=lambda r: nregs[r].address):
+                lines.append(f'case R_{rname}:')
+            lines.append(f'    {vnm} = s->regs[reg];')
+            lines.append('    s->regs[reg] = 0;')
+            lines.append('    break;')
+        else:
+            for rname in sorted(reg_names, key=lambda r: nregs[r].address):
+                lines.append(f'case R_{rname}:')
+                rbm, racc = reg_defs[rname]
+                if bitfield & rbm:
+                    lines.append(f'    {vnm} &= R_{rname}_WMASK;')
+                if racc in (QEMUAccess.RW1C, QEMUAccess.R0W1C):
+                    lines.append(f'    s->regs[reg] &= ~{vnm}; '
+                                 f'/* {racc.upper()} */')
+                elif racc == QEMUAccess.RW0C:
+                    lines.append(f'    s->regs[reg] &= ~{vnm}; /* RW0C */')
```

Comment:
Shouldn't this be:
```suggestion
                    lines.append(f'    s->regs[reg] &= {vnm}; /* RW0C */')
```
Since RW1C is `&= ~mask` and RW0C is `&= mask`?

Reply:
Aside: does this tool handle fields of different R/W access? For example if a register has RW1C & RW0C and fields simultaneously, what happens? And what about read-only fields (should probably guest error on writes to read-only fields)? An example I can think about off the top of my head is the [flash_ctrl](https://opentitan.org/book/hw/top_earlgrey/ip_autogen/flash_ctrl/doc/registers.html#intr_state) or the [i2c](https://opentitan.org/book/hw/ip/i2c/doc/registers.html#intr_state)'s `INTERRUPT_STATE` register.

Reply:
There are not handled. 
I do not think `autoreg.py` and `autotop.py` will ever be able to generate a fully HW-compliant C code, there are too many cases to handle. It should be seen as a helper to avoid manual boring copy/pasting/formatting (maybe AI could do better...) from existing files. The resulting file always needs editing.
I guess combining different R/W access within the same register could be handled, I never encounter this case ... yet :-)